scarlet/operator.py

```diff
@@ -534,122 +505,82 @@ def diagonalsToSparse(diagonals, shape, dtype=np.float64):
-
-        if useNearest:
-            # Only use a single pixel most in line with peak
-            cosNorm = np.zeros_like(cosWeight)
-            columnIndices = np.arange(cosWeight.shape[1])
-            maxIndices = np.argmax(cosWeight, axis=0)
-            indices = maxIndices * cosNorm.shape[1] + columnIndices
-            indices = np.unravel_index(indices, cosNorm.shape)
-            cosNorm[indices] = minGradient
-            # Remove the reference for the peak pixel
-            cosNorm[:, px + py * shape[1]] = 0
-        else:
-            # Normalize the cos weights for each pixel
-            normalize = np.sum(cosWeight, axis=0)
-            normalize[normalize == 0] = 1
-            cosNorm = (cosWeight.T / normalize[:, None]).T
-            cosNorm[mask] = 0
-
-        Cache.set(name, key, cosNorm)
-
-    return cosNorm
+    # Center on the center pixel
+    py, px = int(center[0]), int(center[1])
+    # Calculate the distance between each pixel and the peak
+    x = np.arange(shape[1])
+    y = np.arange(shape[0])
+    X, Y = np.meshgrid(x, y)
+    X = X - px
+    Y = Y - py
+    distance = np.sqrt(X ** 2 + Y ** 2)
 
+    # Find each pixels neighbors further from the peak and mark them as invalid
+    # (to be removed later)
+    distArr, mask = diagonalizeArray(distance, dtype=np.float64)
+    relativeDist = (distance.flatten()[:, None] - distArr.T).T
+    invalidPix = relativeDist <= 0
+
+    # Calculate the angle between each pixel and the x axis, relative to the peak position
+    # (also avoid dividing by zero and set the tan(infinity) pixel values to pi/2 manually)
+    inf = X == 0
+    tX = X.copy()
+    tX[inf] = 1
+    angles = np.arctan2(-Y, -tX)
+    angles[inf & (Y != 0)] = 0.5 * np.pi * np.sign(angles[inf & (Y != 0)])
+
+    # Calculate the angle between each pixel and its neighbors
+    xArr, m = diagonalizeArray(X)
+    yArr, m = diagonalizeArray(Y)
+    dx = (xArr.T - X.flatten()[:, None]).T
+    dy = (yArr.T - Y.flatten()[:, None]).T
+    # Avoid dividing by zero and set the tan(infinity) pixel values to pi/2 manually
+    inf = dx == 0
+    dx[inf] = 1
+    relativeAngles = np.arctan2(dy, dx)
+    relativeAngles[inf & (dy != 0)] = (
+        0.5 * np.pi * np.sign(relativeAngles[inf & (dy != 0)])
+    )
 
-def getRadialMonotonicOp(shape, useNearest=True, minGradient=1, subtract=True):
-    """Create an operator to constrain radial monotonicity
-    This version of the radial monotonicity operator selects all of the pixels closer to the peak
-    for each pixel and weights their flux based on their alignment with a vector from the pixel
-    to the peak. In order to quickly create this using sparse matrices, its construction is a bit opaque.
-    """
-    import scipy.sparse
+    # Find the difference between each pixels angle with the peak
+    # and the relative angles to its neighbors, and take the
+    # cos to find its neighbors weight
+    dAngles = (angles.flatten()[:, None] - relativeAngles.T).T
```

Comment:
I thought from a peak you were only ever looking at the three neighobouring pixels downstream of the light profile. Why then compute weights relative to all the pixels in the image?

Reply:
I'm also deciphering what's going on in this piece...

Reply:
I think that this is a remnant of the way we used to do constraints with ADMM. Looking back on it, as is usually the case looking back at old code, if I was to do it again I would do it differently but the whole thing was built in a few stages.

For Remy, we actually flattened the data into a vector and performed constraints and transforms using matrices so that the linear algebra interpretation was literal. So we needed to build a matrix to apply the constraints and this originally returned a large sparse matrix that we applied to the model with a real dot product. But that was sssslllloooowwww, so we started to create the matrix as a scipy sparse diagonal matrix by converting everything into an array that represented the diagonals and could be used to initialize the sparse matrix. That's the time where the majority of the code comes from, converting data and weights into diagonals for a large sparse matrix. This is probably the last piece of the codebase that hasn't been refactored yet since then (because it always worked).

Looking at the code now for the first time in years, none of this makes sense for what we're doing now, even if we use weights. So fixing it seems beyond the scope of this PR and I'm fine with you just skipping it for now and fixing it as part of a new issue (or assigning it to me to cleanup since it's my mess). But if you want to fix it then you should just probably stop looking at the old code altogether and rewrite it from scratch, since what we're doing now is so much simpler (especially your new flat weights method).

Reply:
+1 I'm ready to throw the towel. I can't even get the nearest-neighbor weighting to work...

Reply:
I got the linear algebra formulation of your operations, I'm just surprised by the size of the matrix, but if it's deprecated then let's make it a ticket and solve it in a later PR.

Reply:
Ah, actually, looking at the code now it isn't doing what I think that you think it's doing.

The `diagonalizeArray` function takes a 2D array, flattens it, and then makes copies of it in the locations of its 8 neighbors (since we don't know a prior which 3 neighbors are closer to the center). So all of the diagonalized arrays are 8xN, where N is the total size of the image. So for each of a pixels 8 neighbors it calculates a relative angle between the pixel and the neighbor, which is what that operation shows, and then a mask is applied to the 5 neighbors that are furthest from the peak.

So this is actually a fairly efficient way of building things up to the line that you highlighted if we restrict ourselves to numpy vectorized operations (I won't say the _most_ efficient way, but it might be close). But since we're using C++ to apply the operator we might as well build it there too, where we can be more efficient for flat weights. But for now if @pmelchior modifies https://github.com/pmelchior/scarlet/blob/flat-mono/scarlet/operator.py#L558-L582 to something like

```
if neighbor_weight == "flat":
    weight = np.zeros(relativeDist.shape, dtype=np.float64)
    weight[~invalidPix] = 1
    return weight
cosWeight = np.cos(dAngles)
cosWeight[invalidPix] = 0
cosWeight[mask] = 0
...
```

that's along the lines of what I was thinking about in my comment during review.

Reply:
That's what I would expect it to do, but I don't see where it says that the size is reduced to the 8 nearest pixels. Also, it seems weird to me to compute the angle relative to those when we know the values these angles will take (or am I missing something?)

Reply:
In https://github.com/pmelchior/scarlet/blob/flat-mono/scarlet/operator.py#L541-L549,
`xArr` and `yArr` are "diagonalized" to be `8xN` pixel arrays. So `dx`, `dy` are also dimension `8xN` and so is `relativeAngles`. So in https://github.com/pmelchior/scarlet/blob/flat-mono/scarlet/operator.py#L557 `relativeAngles` is `8xN` and `angles` is projected to be the same shape.